### ISSUE TITLE
Why do we need to register listeners "immediately after app startup" for Central monitoring

### ISSUE THREAD

**Issue by mr-mmmmore** [781flyingdutchman/background_downloader#180]
Sorry this is not a feature request nor a bug, I have an app that actually relies on _flutter_downloader_ and I would love to migrate it to _background_downloader_, but I need some clarification about the central monitoring process.

It is said in the [readme](https://pub.dev/packages/background_downloader#central-monitoring-and-tracking-in-a-persistent-database) that the registering of the listeners and the calling of `trackTasks` and `resumeFromBackground` should be done _"immediately after app startup"_, why is that, and what would be the consequence of doing it only after performing some other taks at startup?

The tasks I'm talking about are consolidating the download states in the app. As is said at the end of your readme file, _"You should therefore always assume that a task that was started may not complete, and may disappear without providing any status or progress update to indicate why."_ This is why in my current app I have a process that checks the download states and takes decisions about relaunching some downloads, cancelling some others, cleaning files and download states, etc. This process can't be done while a queue is active otherwise it might mess the whole thing. So in my app I call the `FlutterDownloader.initialize` method and register the isolate callback only after that this job is done.

I would like to do the same with your plugin but you seem to say that I shouldn't wait to initialize it after the app startup, so could you explain why and maybe if there is a way to achieve what I want to do?

**Comment by 781flyingdutchman** [comment]
You can think of your app and a background (download) task as two independent processes, that do not have to be active at the same time. So, your background task can run without your app running.  The moment your app starts, Flutter creates the connection to the background_downloader plugin, and the plugin will then feed status or progress updates to your app. To ensure consistent state you therefore need to start listening to those updates immediately, as otherwise you may lose some, resulting in incorrect state.  The `trackTasks` method intercepts updates to ensure consistent state, so that is the first one you want to activate, and do so immediately (you can even do this before starting your own listener) so that you're not missing updates on currently active tasks.  Next, `resumeFromBackground` will fetch all status and progress updates that may have happened while your app was stopped, so again to ensure consistent state you likely want to have those processed before you do anything else (and you can choose to do this before registering listeners, just realize that you are then relying on the tracked state).  Next, if you want to do anything in your app in response to status or progress changes, you want to register your own listeners.  If you are using task tracking then you can in theory delay registering listeners until you really need them, but in general it is best to establish everything upfront in a singleton type object, where you keep state (possibly with support from the tracking functionality) and where you are always listening to updates, as you never know when they may happen.  Because of that, the typical order is 1) register listeners, 2) activate task tracking, 3) call `resumeFromBackground`.  All current and 'missed' updates will then hit your listener, as well as the task tracking mechanism.

My suggestion would be to follow that approach (i.e. register everything immediately after startup) and to call a 'cleanup' function a second or so after startup where you do your checks and relaunches etc.  In the app that I use background_downloader in, I check for files that exist (and update download state to 'downloaded' if they do), then I compare downloads in progress with the ones I have marked as 'must download' and I reschedule the ones that must be downloaded but are somehow not in progress.  This approach works for me and has not led to any issues.  I use a separate task tracking mechanism (not the built-in one) but that shouldn't make a difference.

Hope this helps, let me know if you have any other questions.

**Comment by mr-mmmmore** [comment]
Thank you for your detailed explanation!

It sound like it may be possible to simplify a lot the download logic in my app by relying mostly on your plugin, if so this is great news 🎉

What I understand is that the plugin kinda "ensures" a consistent state of the download process, and that by listening to the events asap I can get all the events that allows the app to maintain its state consistent as well, is that right?

If so, I'd be interested to know: what are the cases where you still have to do some cleaning in your app:
- the download is in error status,
- the  status is final but the file is not there,
- what else?

**Comment by 781flyingdutchman** [comment]
The only other one - I think - is "I've marked and enqueued the file for download, but isn't active, i.e. isn't returned by `allTasks` but its status is not a final state"

**Comment by 781flyingdutchman** [comment]
Depending on how complicated your download state management is, you may also want to consider using `SqlitePersistentStorage` instead of the default, which uses a simple filesystem based storage. If you do, your data is stored in a simple SQLite table that you can also query from within Dart, and you could add additional columns etc if you need to. Take a look at the migration documentation for inspiration.

**Comment by mr-mmmmore** [comment]
OK, I think that I won't need to query the DB but I keep this possibility in mind.

Thank you for your help and your responses, you can close the ticket.
